### PR TITLE
Add support for structures and enumerations.

### PR DESCRIPTION
**Pull request by KlimMixer**
Fixes: #266

### CODE REVIEW

File: src/source.ts [hunk 1]
```diff
@@ -251,7 +287,7 @@ abstract class SourceContractItem implements Linkable {
   @memoize
   get args(): SourceTypedVariable[] {
     return SourceTypedVariableArray.fromParameterList(
-      this.astNode.parameters,
+      ('parameters' in this.astNode)?this.astNode.parameters:{parameters: []},
```
Comment (frangio) [inline src/source.ts:294]: I'd prefer to rollback this line, and instead change the type of `astNode` to be `FunctionDefinition | EventDefinition | ModifierDefinition`.

File: src/source.ts [hunk 2]
```diff
@@ -261,7 +297,7 @@ abstract class SourceContractItem implements Linkable {
 
   @memoize
   get natspec(): NatSpec {
-    if (this.astNode.documentation === null || this.astNode.documentation === undefined) {
+    if (!('documentation' in this.astNode) || this.astNode.documentation === null || this.astNode.documentation === undefined) {
```
Comment (frangio) [inline src/source.ts:304]: This should not be necessary. We should add the `documentation` member in the new interfaces.

File: src/source.ts [hunk 3]
```diff
@@ -177,13 +185,17 @@ export class SourceContract implements Linkable {
     const functions = groupBy(this.functions, f => f.contract.astId);
     const events = groupBy(this.events, f => f.contract.astId);
     const modifiers = groupBy(this.modifiers, f => f.contract.astId);
+    const structs = groupBy(this.enums, f => f.contract.astId);
```
Comment (frangio) [inline src/source.ts:188]:
```suggestion
    const structs = groupBy(this.structs, f => f.contract.astId);
```

**Review by frangio — COMMENTED** [review]
Thanks @KlimMixer!

**Comment by KlimMixer** [comment]
@frangio Thank you. Updated, and fixed all problems.

**Comment by frangio** [comment]
I was wrong about documentation, these nodes don't have documentation in the AST. I also decided to simplify things a little, enum and struct members don't need to be Linkable. Please take a look and let me know if you have any feedback.

**Comment by KlimMixer** [comment]
Thank you for the simplification and nice realization of differentiation between functions events modifiers and other contract items.
And next, I will write tests and push for the review.

**Comment by frangio** [comment]
For testing you should make sure there are structs and enums in here for example: https://github.com/OpenZeppelin/solidity-docgen/tree/master/fixtures/basic/input

**Comment by frangio** [comment]
Can you create a single new fixture with both struct and enum? And give it a custom template (see how `custom` does it) that specifically tests struct and enum and its members.

**Review by frangio — DISMISSED** [review]
Looks good. Thank oyu @KlimMixer!

**Review by frangio — APPROVED** [review]

**Pull request merged**